### PR TITLE
Resource changes

### PR DESCRIPTION
**Pull request by andrew-mestas**
The content to be added to Changelog is as follows:
* https://github.com/Azure/azure-xplat-cli/issues/2751
* https://github.com/Azure/azure-xplat-cli/issues/2689
* https://github.com/Azure/azure-xplat-cli/issues/2552

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@andrew-mestas__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (v-anmest). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: lib/commands/arm/armsdk/resourceManagementClient.js [hunk 1]
```diff
@@ -5934,6 +5934,7 @@ var ResourceOperations = ( /** @lends ResourceOperations */ function() {
     });
   };
   
+
```
Comment (TianoMS) [inline lib/commands/arm/armsdk/resourceManagementClient.js]: Don't need this new line.

File: lib/plugins.arm.json [hunk 2]
```diff
@@ -65692,6 +65700,23 @@
               "short": "-p",
               "long": "--properties",
               "description": "a JSON-formatted string containing properties"
+            },            
+            {
+              "flags": "-s --sku <sku>",
+              "required": -17,
```
Comment (TianoMS) [inline lib/plugins.arm.json]: What's the meaning of -17 here?

File: lib/plugins.arm.json [hunk 3]
```diff
@@ -65716,7 +65741,7 @@
           "name": "set",
           "description": "Updates a resource in a resource group without any templates or parameters",
           "fullName": "resource set",
-          "usage": "[options] <resource-group> <name> <resource-type> <properties> <api-version>",
```
Comment (TianoMS) [inline lib/plugins.arm.json]: why is "properties" removed?

File: lib/commands/arm/group/group._js [hunk 4]
```diff
@@ -167,7 +167,6 @@ exports.init = function (cli) {
         }
       }
     });
-
```
Comment (TianoMS) [inline lib/commands/arm/group/group._js]: Don't need this change.

File: lib/commands/arm/resource/resource._js [hunk 5]
```diff
@@ -93,32 +99,46 @@ exports.init = function (cli) {
...
+      var resourceLocation = location || resource.location;
+      if (!resourceLocation){
+        cli.missingArgument('location');
+      }
+      var skuObj = options.sku;
+      var planObj = options.plan;
+      var sku = {};
+      if (skuObj) {
+        sku = jsonlint.parse(skuObj);
+        requestProperties.sku = sku;
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: move this line out of "if" ?

File: lib/commands/arm/resource/resource._js [hunk 6]
```diff
@@ -93,32 +99,46 @@ exports.init = function (cli) {
...
+      }
+      var plan = {};
+      if (planObj) {
+        plan = jsonlint.parse(planObj);
+        requestProperties.plan = plan;
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: move this line out of "if" ?

File: lib/commands/arm/resource/resource._js [hunk 7]
```diff
@@ -391,26 +389,118 @@ exports.init = function (cli) {
...
+        client.resourceGroups.deleteMethod(name, _);
+      } finally {
+        progress.end();
+      }
+    }
+  }
+  function extractAllValues(resourceId){
+    var values = {};
+    var array = resourceId.split('/');
+    array = array.slice(array.indexOf('subscriptions'), array.length)
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: what's the design here? What happens if array.indexOf('subscriptions') returns -1?

File: lib/commands/arm/resource/resource._js [hunk 8]
```diff
@@ -391,26 +389,118 @@ exports.init = function (cli) {
...
+      try {
+        client.resourceGroups.deleteMethod(name, _);
+      } finally {
+        progress.end();
+      }
+    }
+  }
+  function extractAllValues(resourceId){
+    var values = {};
+    var array = resourceId.split('/');
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: give a more meaningful name, e.g. "segments".

File: lib/commands/arm/resource/resource._js [hunk 9]
```diff
@@ -46,6 +46,9 @@ exports.init = function (cli) {
     .option('-o --api-version <api-version>', $('the API version of the resource provider'))
     .option('--parent <parent>', $('the name of the parent resource (if needed), in path/path/path format'))
     .option('-p --properties <properties>', $('a JSON-formatted string containing properties'))
+    .option('-s --sku <sku>', $('a JSON-formatted string containing sku properties'))
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: -s is reserved switch for subscription. Please use something else.

File: lib/commands/arm/resource/resource._js [hunk 10]
```diff
@@ -61,6 +64,9 @@ exports.init = function (cli) {
     .option('-r --resource-type <resource-type>', $('the resource type'))
     .option('-p --properties <properties>', $('a JSON-formatted string containing properties'))
     .option('-o --api-version <api-version>', $('the API version of the resource provider'))
+    .option('-s --sku <sku>', $('a JSON-formatted string containing sku properties'))
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: same as above.

File: lib/commands/arm/resource/resource._js [hunk 11]
```diff
@@ -46,6 +46,9 @@ exports.init = function (cli) {
     .option('-o --api-version <api-version>', $('the API version of the resource provider'))
     .option('--parent <parent>', $('the name of the parent resource (if needed), in path/path/path format'))
     .option('-p --properties <properties>', $('a JSON-formatted string containing properties'))
+    .option('-s --sku <sku>', $('a JSON-formatted string containing sku properties'))
+    .option('--plan <plan>', $('a JSON-formatted string containing plan properties'))
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: It is very cumbersome to provide a JSON formatted string as one needs to escape quotes `"{\"foo\":\"bar\"}"`. And if the  object is large then it provides a bad exprience to the end user. It would be nice if you also provide an option where the user can provide the json object in a file. Both the options should be mutually exclusive. Thus the user can decide what he wants to do.

File: lib/commands/arm/resource/resource._js [hunk 12]
```diff
@@ -46,6 +46,9 @@ exports.init = function (cli) {
...
+    .option('--isFullObject', $('Indicates that the property object includes other options such as location, tags, sku, and/or plan.'))
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: I dont like this switch name `--isFullObject`. A better name would be `--expand` or `--details`. The help text that you have is correct. It provides enough info to the user.

File: lib/commands/arm/resource/resource._js [hunk 13]
```diff
@@ -93,32 +99,46 @@ exports.init = function (cli) {
...
+      var skuObj = options.sku;
+      var planObj = options.plan;
+      var sku = {};
+      if (skuObj) {
+        sku = jsonlint.parse(skuObj);
+        requestProperties.sku = sku;
+      }
+      var plan = {};
+      if (planObj) {
+        plan = jsonlint.parse(planObj);
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: JSON.parse throws an error if something goes wrong. I assume that jsonlint.parse() would do the same. Please put this statement in try catch block so that you can provide a meaningful error message along with the original error.

File: lib/commands/arm/resource/resource._js [hunk 14]
```diff
@@ -228,32 +245,30 @@ exports.init = function (cli) {
       }
     });
 
-  resource.command('show [resource-group] [name] [resource-type] [api-version]')
+  resource.command('show [resource-group] [name] [resource-type] [api-version] [resource-id]')
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: I am not sure if one more positional parameter is useful. It always confuses the user. I think we would be better off by just having it as a switch parameter

File: lib/commands/arm/resource/resource._js [hunk 15]
```diff
@@ -275,97 +289,80 @@ exports.init = function (cli) {
...
         progress.end();
       }
-
       cli.interaction.formatOutput(resource, function (resource) {
         showResource(resource, true);
       });
     });
 
-  resource.command('delete [resource-group] [name] [resource-type] [api-version]')
+  resource.command('delete [resource-group] [name] [resource-type] [api-version] [resource-id]')
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: same comment as above. Actually resource-id and (resource-group, name, resource-type, parent) should be mutually exclusive. Resource ID is made up of all those things. It makes no sense to ask the user to provide both. You should make sure that both are mutually exclusive and resource-id should definitely not be a positional parameter
/cc @TianoMS - What do you think?

File: lib/commands/arm/resource/resource._js [hunk 16]
```diff
@@ -391,26 +389,118 @@ exports.init = function (cli) {
...
+      return cli.missingArgument('apiVersion');
+    }
+  }
+
+  function validateApiVersion(dateString) {
+    var dateFormat = /^(19|20)\d\d([- /.])(0[1-9]|1[012])\2(0[1-9]|[12][0-9]|3[01])/;
+    if (dateFormat.exec(dateString)) {
+      return dateString;
+    } else {
+      console.log('');
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: please dont use console.log statements. It corrupts the output when someone uses --json format.

File: lib/util/utils.js [hunk 17]
```diff
@@ -117,12 +117,15 @@ exports.createAutoRestClient = function (FactoryMethod, subscription, options) {
   
   // there are a subset of clients that do not require a subscription
   // TODO: need to figure out a way to pass options into factoryMethodOptions.
-  if(options && options.noSubscription) {
-    return new FactoryMethod(credentials, factoryMethodOptions);
-  }
-  else {
+  if (options && options.noSubscription) {
```
Comment (amarzavery) [inline lib/util/utils.js]: Why was this change required?

**Comment by andrew-mestas** [comment]
Majority of comments addressed in c465db8.

**Comment by amarzavery** [comment]
please resolve conflicts

**Comment by amarzavery** [comment]
please address other comments that were not addressed in the previous commit

File: lib/util/utils.js (hunk 17 above)
Comment (andrew-mestas) [inline lib/util/utils.js]: Merged from location list changes.

**Comment by amarzavery** [comment]
@andrew-mestas - CI failing

File: test/commands/arm/resource/arm.resource-tests.js [hunk 18]
```diff
@@ -89,7 +89,92 @@ describe('arm', function () {
           });
         });
       });
-      
+      it('should work with sku', function (done) {
```
Comment (vivsriaus) [inline test/commands/arm/resource/arm.resource-tests.js]: nit: please add a newline after each test for better readability

File: lib/commands/arm/resource/resource._js [hunk 19]
```diff
@@ -46,6 +46,9 @@ exports.init = function (cli) {
     .option('-o --api-version <api-version>', $('the API version of the resource provider'))
     .option('--parent <parent>', $('the name of the parent resource (if needed), in path/path/path format'))
     .option('-p --properties <properties>', $('a JSON-formatted string containing properties'))
+    .option('--sku <sku>', $('a JSON-formatted string containing sku properties'))
```
Comment (vivsriaus) [inline lib/commands/arm/resource/resource._js]: Can we please make location an optional parameter? Location is not required for extension resource types like resource links

File: lib/commands/arm/resource/resource._js [hunk 20]
```diff
@@ -61,6 +64,9 @@ exports.init = function (cli) {
     .option('-r --resource-type <resource-type>', $('the resource type'))
     .option('-p --properties <properties>', $('a JSON-formatted string containing properties'))
     .option('-o --api-version <api-version>', $('the API version of the resource provider'))
+    .option('--sku <sku>', $('a JSON-formatted string containing sku properties'))
+    .option('--plan <plan>', $('a JSON-formatted string containing plan properties'))
```
Comment (vivsriaus) [inline lib/commands/arm/resource/resource._js]: We need resourceId for set resource as well.

File: lib/commands/arm/resource/resource._js [hunk 21]
```diff
@@ -93,32 +99,40 @@ exports.init = function (cli) {
...
     var properties = {};
     if (propertiesParam) {
-      properties = jsonlint.parse(propertiesParam);
+      properties = parsed(propertiesParam, 'properties');
+    }
+    if(!options.isFullObject){
+      var tags = {};
+      tags = tagUtils.buildTagsParameter(tags, options);
+      var resourceLocation = location || resource.location;
+      if (!resourceLocation){
```
Comment (vivsriaus) [inline lib/commands/arm/resource/resource._js]: Same comment as above. Location shouldn't be a required parameter when creating extension resource types like links

File: lib/commands/arm/resource/resource._js [hunk 22]
```diff
@@ -234,26 +245,25 @@ exports.init = function (cli) {
...
-      } else if (!name) {
-        return cli.missingArgument('name');
-      } else if (!resourceType) {
-        return cli.missingArgument('resourceType');
-      } else if (!apiVersion) {
-        return cli.missingArgument('apiVersion');
+      if(!options.resourceId) {
+        checkMissingParameters(resourceGroup, name, resourceType, apiVersion);
+      } else {
+        var values = extractAllValues(options.resourceId);
```
Comment (vivsriaus) [inline lib/commands/arm/resource/resource._js]: @amarzavery How do we enable something like a parameter set for CLI commands? As in, for show resource command, I can identify a resource either by:
a. fully qualified resource id
or
b. combination of resource name, resource type and resource group

With the changes here, I can specify both of the above, but the code doesn't handle that. Should we add a crude if statement, saying "please specify either resource id or name/type/group. Not both"?

File: lib/commands/arm/resource/resource._js [hunk 23]
```diff
@@ -287,85 +295,69 @@ exports.init = function (cli) {
...
-          identity.resourceProviderNamespace,
-          identity.parentResourcePath,
-          identity.resourceType,
-          identity.resourceName,
-          apiVersion, _);
-      } finally {
-        progress.end();
+      if(resourceType === undefined && name === undefined) {
+        if (!resourceGroup) { return cli.missingArgument('resourceGroup'); }
+        deleteResourceGroup(resourceGroup, options, client, progress, _);
```
Comment (vivsriaus) [inline lib/commands/arm/resource/resource._js]: Why are we calling delete resource group from a delete resource command?

File: lib/commands/arm/resource/resource._js [hunk 24]
```diff
@@ -391,26 +384,124 @@ exports.init = function (cli) {
...
+        resourceName: name,
+        resourceProviderNamespace: resourceUtils.getProviderName(resourceType),
+        resourceProviderApiVersion: apiVersion,
+        resourceType: resourceUtils.getResourceTypeName(resourceType),
+        // TODO: parent should be optional in the API. temporary workaround.
+        parentResourcePath: __.isString(options.parent) ? options.parent : ''
+      };
+
+      try {
+        client.resources.get(resourceGroup,
```
Comment (vivsriaus) [inline lib/commands/arm/resource/resource._js]: Use check existence instead

File: lib/commands/arm/armsdk/resourceManagementClient.js [hunk 25]
```diff
@@ -5850,7 +5850,7 @@ var ResourceOperations = ( /** @lends ResourceOperations */ function() {
     url2 = url2 + encodeURIComponent(sourceResourceGroupName);
     url2 = url2 + '/moveResources';
     var queryParameters = [];
-    queryParameters.push('api-version=2015-11-01');
+    queryParameters.push('api-version=2015-11-01');
```
Comment (amarzavery) [inline lib/commands/arm/armsdk/resourceManagementClient.js]: don't need this change either

File: lib/commands/arm/resource/resource._js [hunk 26]
```diff
@@ -391,26 +384,124 @@ exports.init = function (cli) {
...
+        identity.parentResourcePath,
+        identity.resourceType,
+        identity.resourceName,
+        apiVersion, _);
+    } finally {
+      progress.end();
+    }
+  } 
+
+  function deleteResourceGroup(name, options, client, progress, _){
```
Comment (vivsriaus) [inline lib/commands/arm/resource/resource._js]: This shouldn't be here. We have a separate delete resource group command

File: test/commands/arm/resource/arm.resource-tests.js [hunk 27]
```diff
@@ -401,46 +569,120 @@ describe('arm', function () {
...
+
+                suite.execute('group delete %s --quiet --json', groupName, function () {
+                  done();
+                });
+              });
+            });
+          });
+        });
+      });
+    });
```
Comment (vivsriaus) [inline test/commands/arm/resource/arm.resource-tests.js]: Can you add a test for [this](https://github.com/Azure/azure-xplat-cli/issues/2552)?

File: lib/commands/arm/resource/resource._js (hunk 22 above)
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: We don't have concept of a parameter set. We solve this by explicitly mentioning this in the help and making a check to findout that both are not provided. If they are then we throw letting the user know that blah blah paramaters are mutually exclusive.

File: lib/commands/arm/group/group._js (hunk 4 above)
Comment (TianoMS) [inline lib/commands/arm/group/group._js]: @andrew-mestas seems you missed this one. We don't need this change.

File: lib/commands/arm/resource/resource._js [hunk 28]
```diff
@@ -93,32 +99,40 @@ exports.init = function (cli) {
       parentResourcePath: __.isString(options.parent) ? options.parent : ''
     };
 
-    var resource = withProgress(util.format($('Getting resource %s'), name),
+    var resource = withProgress(util.format($('Getting resource %s'), name),
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: remove the extra spaces at the end of line.

File: lib/commands/arm/resource/resource._js [hunk 29]
```diff
@@ -391,26 +384,124 @@ exports.init = function (cli) {
...
         progress.end();
       }
-
     });
 
+  function parsed(obj, type){
+    try{
+      return jsonlint.parse(obj);
+    } catch(e){
+      throw new Error('Error with ' + type + ' object. ' + e);
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: I think we can say something like: "Error when parsing the input value for '*type*': '*exception*'".

File: lib/commands/arm/resource/resource._js [hunk 30]
```diff
@@ -287,85 +295,69 @@ exports.init = function (cli) {
...
-
       if (!options.quiet && !cli.interaction.confirm(util.format($('Move selected resources in %s to %s? [y/n] '), sourceGroup, destinationGroup), _)) {
         return;
       }
-
       var subscription = profile.current.getSubscription(options.subscription);
       var client = utils.createResourceClient(subscription);
+      apiVersion = validateApiVersion(apiVersion);
+      client.apiVersion = apiVersion ? apiVersion : client.apiVersion;
+      log.info('Api Version: ', client.apiVersion);
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: do we need this log?

File: lib/commands/arm/resource/resource._js [hunk 31]
```diff
@@ -391,26 +384,124 @@ exports.init = function (cli) {
...
+        client.resourceGroups.deleteMethod(name, _);
+      } finally {
+        progress.end();
+      }
+    }
+  }
+  function extractAllValues(resourceId){
+    var values = {};
+    var segments = resourceId.split('/');
+    var firstValue = segments.indexOf('subscriptions') === -1 ? 'resourceGroups' : 'subscriptions';
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: Are we assuming the first segment is either "subscriptions" or "resourcegroups"? This might not be correct.

File: lib/commands/arm/resource/resource._js [hunk 32]
```diff
@@ -391,26 +384,124 @@ exports.init = function (cli) {
...
+    segments = segments.slice(segments.indexOf(firstValue), segments.length)
+                 .filter(function(element){ 
+                   return element !== '';
+                 });
+    var providerList = segments.slice(segments.indexOf('providers') + 1, segments.length-1);
+      values['subscription'] = firstValue === 'subscriptions' ? segments[segments.indexOf('subscriptions') + 1] : 'No subscription provided.';
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: when there's no "subscriptions" segment,  don't set value['subscription'].

File: lib/commands/arm/resource/resource._js [hunk 33]
```diff
@@ -391,26 +384,124 @@ exports.init = function (cli) {
...
+  }
+  function extractAllValues(resourceId){
+    var values = {};
+    var segments = resourceId.split('/');
+    var firstValue = segments.indexOf('subscriptions') === -1 ? 'resourceGroups' : 'subscriptions';
+    segments = segments.slice(segments.indexOf(firstValue), segments.length)
+                 .filter(function(element){ 
+                   return element !== '';
+                 });
+    var providerList = segments.slice(segments.indexOf('providers') + 1, segments.length-1);
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: what is providerList here? Is it something like "providers, Microsoft.Web, sites, mywebsite"?

File: lib/commands/arm/resource/resource._js [hunk 34]
```diff
@@ -391,26 +384,124 @@ exports.init = function (cli) {
...
+      if(segments.length > 4) {
+        values['providerList'] = providerList.join('/');
+        values['provider'] = providerList[0];
+        values['resourceType'] = providerList.slice(1, providerList.length).join('/');
+        values['name'] = segments[segments.length - 1]; 
+      }
+    return values;
+  }
+
+  function checkMissingParameters(resourceGroup, name, resourceType, apiVersion){
```
Comment (TianoMS) [inline lib/commands/arm/resource/resource._js]: can we rename "name" to "resourceName"?

File: lib/plugins.arm.json [hunk 35]
```diff
@@ -66080,6 +66128,15 @@
               "description": "the resource type"
             },
             {
+              "flags": "-i --resource-id <resource-id>",
+              "required": -20,
```
Comment (TianoMS) [inline lib/plugins.arm.json]: what does -20 mean here?

File: test/data/zone_file_export.txt [hunk 36]
```diff
@@ -1,13 +1,13 @@
 ; Exported zone file from Azure DNS
 ; Resource Group Name: xplat-test-dns-zone
 ; Zone name: example1.com
-; Date and time (UTC): Tue Jun 28 2016 00:29:58 GMT+0000
+; Date and time (UTC): Tue Aug 23 2016 21:41:29 GMT+0000
```
Comment (TianoMS) [inline test/data/zone_file_export.txt]: Do we need this change?

**Comment by amarzavery** [comment]
@andrew-mestas - resolve merge conflicts and resolve all the comments for this PR to be merged in this release. Cut off is today 5.00 PM PST

File: lib/plugins.arm.json (hunk 35 above)
Comment (andrew-mestas) [inline lib/plugins.arm.json]: Fixed it should be zero to show optional

File: test/data/zone_file_export.txt (hunk 36 above)
Comment (andrew-mestas) [inline test/data/zone_file_export.txt]: Not sure when this happened just pulled and fixed merge conflict

File: lib/commands/arm/resource/resource._js (hunk 30 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: used it to debug just removed

File: lib/commands/arm/resource/resource._js (hunk 29 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: Just added

File: lib/commands/arm/resource/resource._js (hunk 23 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: Took it out

File: lib/commands/arm/resource/resource._js (hunk 20 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: Fixed

File: lib/commands/arm/resource/resource._js (hunk 19 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: Fixed

File: lib/commands/arm/resource/resource._js (hunk 32 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: Tenant level will submit an issue

**Comment by andrew-mestas** [comment]
@amarzavery Will this PR be merged?

File: lib/commands/arm/resource/resource._js [hunk 37]
```diff
@@ -61,29 +63,24 @@ exports.init = function (cli) {
     .option('-r --resource-type <resource-type>', $('the resource type'))
     .option('-p --properties <properties>', $('a JSON-formatted string containing properties'))
     .option('-o --api-version <api-version>', $('the API version of the resource provider'))
+    .option('-i --resource-id <resource-id>', $('the resource id'))
+    .option('--sku <sku>', $('a JSON-formatted string containing sku properties'))
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: An example and a link to find all the possible values for sku and plan for --sku and --plan would be really helpful for the customer

File: lib/commands/arm/resource/resource._js [hunk 38]
```diff
@@ -234,48 +235,36 @@ exports.init = function (cli) {
     .option('-g --resource-group <resource-group>', $('the resource group name'))
     .option('-n --name <name>', $('the resource name'))
     .option('-r --resource-type <resource-type>', $('the resource type'))
+    .option('-i --resource-id <resource-id>', $('the resource id'))
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: General comment for all the commands that accept resource-id as the parameter

> Please specify in the help that resource-id and (resource-group, name, resource-type, parent) are mutually exclusive.

File: lib/commands/arm/resource/resource._js [hunk 39]
```diff
@@ -287,85 +276,49 @@ exports.init = function (cli) {
...
       var sourceGroup = resourceUtils.getSourceResourceGroupForMove(resources);
-
       if (!options.quiet && !cli.interaction.confirm(util.format($('Move selected resources in %s to %s? [y/n] '), sourceGroup, destinationGroup), _)) {
         return;
       }
-
       var subscription = profile.current.getSubscription(options.subscription);
       var client = utils.createResourceClient(subscription);
+      apiVersion = validateApiVersion(apiVersion);
+      client.apiVersion = apiVersion ? apiVersion : client.apiVersion;
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: @vivsriaus @TianoMS @andrew-mestas 

The api-version of the individual resource type may be completely different.
Isn't this weird? IMO, it would be best to throw an error rather than taking the api-version of the resource manager api.

File: lib/commands/arm/resource/resource._js [hunk 40]
```diff
@@ -391,26 +345,131 @@ exports.init = function (cli) {
...
+      return cli.missingArgument('apiVersion');
+    }
+  }
+
+  function validateApiVersion(dateString) {
+    var dateFormat = /^(19|20)\d\d([- /.])(0[1-9]|1[012])\2(0[1-9]|[12][0-9]|3[01])/;
+    if (dateFormat.exec(dateString)) {
+      return dateString;
+    } else {
+      log.error('Incorrect Format. Please enter in the following format (YYYY-MM-DD). Defaulting to 2016-02-01');
```
Comment (amarzavery) [inline lib/commands/arm/resource/resource._js]: If you are proceeding ahead then this should be logged as a warning.
If you intend to halt then it should be logged as an error and the error should be thrown.

File: test/commands/arm/resource/arm.resource-tests.js [hunk 41]
```diff
@@ -35,7 +35,7 @@ var createdGroups = [];
 var createdResources = [];
 
 describe('arm', function () {
-  describe('resource', function () {
+  describe.only('resource', function () {
```
Comment (amarzavery) [inline test/commands/arm/resource/arm.resource-tests.js]: please remove this only. This will make mocha run only the resource test suite.

**Comment by amarzavery** [comment]
@andrew-mestas @vivsriaus @TianoMS - Please respond to my comments for this to be merged in time for sept 21 release

File: lib/commands/arm/resource/resource._js (hunk 37 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: I agree I'll see about adding that

File: test/commands/arm/resource/arm.resource-tests.js (hunk 41 above)
Comment (andrew-mestas) [inline test/commands/arm/resource/arm.resource-tests.js]: Ok

File: lib/commands/arm/resource/resource._js (hunk 40 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: sounds good

File: lib/commands/arm/resource/resource._js (hunk 39 above)
Comment (andrew-mestas) [inline lib/commands/arm/resource/resource._js]: Will discuss

Comment (vivsriaus) [inline lib/commands/arm/resource/resource._js]: @andrew-mestas @amarzavery I agree - api-version should be a required parameter for resource CRUD commands

Comment (yugangw-msft) [inline lib/commands/arm/resource/resource._js]: Can we get from the latest api-version through resource provider `get' operation?  Asking for less is always good for the field which is hard for people to come up with

**Comment by amarzavery** [comment]
@andrew-mestas @vivsriaus @TianoMS - Please send an update for this to be merged in time for sept 21 release

**Comment by amarzavery** [comment]
@andrew-mestas @vivsriaus @TianoMS- 2 more days remaining, if you want to get this in. Else the change will be in november release.
